mainauto.c

```diff
@@ -0,0 +1,39 @@
+#include <stdio.h>
+#include "display.h"
+#include "mapio.h"
+#include "datastrc.h"
+#include "control.h"
+
+int main(int argc, char &argv){
+
+allocate_memory(*gmdt);
+//First allocate some memory
+
+allocate_pngns(*gmdt);
+//allocate memory for penguins
+
+read_data(argv[1], *gmdt);
+
+if(gmdt->crnt_player == 0){
+    generate_map(*gmdt);
+    crnt_player = 2;
+    write_data(*gmdt);
+    return 1;
+}
+
+if(gmdt->phase == 1){
+    place_penguin(*gmdt);
+
+    //update phase
+
+    write_data(*gmdt);
+    return 1;
+}
+
+if(gmdt->phase == 2){
+    move_player(*gmdt);
+    //move_player functions update current player number
+    write_data(*gmdt)
+    return 1;
+
+}
```

Comment:
This is good, but just to indicate that it's the AI doing the moves, call the functions `place_penguin_ai` and `move_player_ai`.

Of course change `*gmdt` to `&gmdt` as well.

`write_data` was a correct name, sorry for the confusion, but it was changed today to `save_file`, and it now takes two arguments - the name of the output file (which should be stored as argv[2] here), and a pointer to gmdt. So change that as well, please.

Just before the return and right after the data save, add some **free memory** functions. I've added some to the datastrc.c file - they work similarly to the memory allocation functions, but in reverse. So basically do what you've done at the beginning and what I've told you to remove, just with the **free** functions instead of the **allocate** ones.

Don't return 1, return 0 instead. In the main() function, 1 means an error, and 0 means successful execution of the program.

And last thing - you're missing a '}' at the very end, to close the `int main(){`.